File: Leetcode刷题（含有笔记）/hot100/两两交换链表中的节点.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode pre = dummy;

        if (head == null || head.next == null) {
            return head;
        }

        while (pre.next != null && pre.next.next != null) {
            ListNode first = pre.next;
            ListNode second = pre.next.next;

            // Swap nodes
            first.next = second.next;
            second.next = first;
            pre.next = second;

            // Move pre to the next pair
            pre = pre.next.next;
        }

        return dummy.next;
    }
}

File: Go刷题/两两交换链表中的节点.go
func swapPairs(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	if head.Next == nil {
	    return head
	}
	dummy := &ListNode{Next: head}
	pre : =dummy
	while pre.Next != nil && pre.Next.Next != nil {
	    temp := pre.Next
		temp2 := pre.Next.Next
		pre.Next = temp.Next
		temp.Next = temp2.Next
		temp2.Next = temp
		pre = temp//很重要
	}
	return dummy.Next
}